logger/logSubsystem.go

```diff
@@ -0,0 +1,110 @@
+var loggers map[string]*logger
+var defaultLogOut LogOutputHandler
+var defaultLogLevel = LogInfo
+
+func init() {
+	logMut.Lock()
+	defer logMut.Unlock()
+
+	loggers = make(map[string]*logger)
+	defaultLogOut = &logOutputSubject{}
+	_ = defaultLogOut.AddObserver(os.Stdout, ConsoleFormatter{})
+}
+
+// Get returns a log based on the name provided, generating a new log if there is no log with provided name
+func Get(name string) *logger {
+	logMut.Lock()
+	defer logMut.Unlock()
+
+	logger, ok := loggers[name]
+	if !ok {
+		logger = newLogger(name, defaultLogLevel, defaultLogOut)
+		loggers[name] = logger
+	}
+
+	return logger
+}
+
+// ConvertHash generates a short-hand of provided bytes slice showing only the first 3 and the last 3 bytes as hex
+// in total, the resulting string is maximum 13 characters long
+func ConvertHash(hash []byte) string {
+	if len(hash) == 0 {
+		return ""
+	}
+	if len(hash) < 6 {
+		return hex.EncodeToString(hash)
+	}
+
+	prefix := hex.EncodeToString(hash[:3])
+	suffix := hex.EncodeToString(hash[len(hash)-3:])
+	return prefix + ellipsisCharacter + suffix
+}
+
+// SetLogLevel changes the log level of the contained loggers. The expected format is "LOG_LEVEL|MATCHING_STRING".
+// If matching string is * it will change the log levels of all contained loggers and will also set the
+// defaultLogLevelProperty. Otherwise, the log level will be modified only on those loggers that will contain the
+// matching string on any position.
+// For example, having the parameter "DEBUG|process" will set the DEBUG level on all loggers that will contain
+// the "process" string in their name ("process/sync", "process/interceptors", "process" and so on).
+func SetLogLevel(logLevelAndPattern string) error {
+	logLevel, pattern, err := ParseLogLevelAndMatchingString(logLevelAndPattern)
+	if err != nil {
+		return err
+	}
+
+	logMut.Lock()
+	defer logMut.Unlock()
+
+	setLogLevelOnMap(loggers, pattern, logLevel)
+	setLogLevelToVar(&defaultLogLevel, pattern, logLevel)
+
+	return nil
+}
+
+// AddLogObserver adds a new observer (writer + formatter) to the already built-in log observers queue
+// This method is useful when adding a new output device for logs is needed (such as files, streams, API routes and so on)
+func AddLogObserver(w io.Writer, formatter Formatter) error {
+	return defaultLogOut.AddObserver(w, formatter)
+}
+
+func setLogLevelOnMap(loggers map[string]*logger, pattern string, logLevel LogLevel) {
+	for name, log := range loggers {
+		isMatching := pattern == "*" || strings.Contains(name, pattern)
+		if isMatching {
+			log.SetLevel(logLevel)
+		}
+	}
+}
+
+func setLogLevelToVar(dest *LogLevel, pattern string, logLevel LogLevel) {
+	if pattern == "*" {
+		*dest = logLevel
+	}
+}
+
+// ParseLogLevelAndMatchingString can parse a string in the form "LOG_LEVEL|MATCHING_STRING" into its
+// corresponding log level and matching string. Errors if something goes wrong.
+// Example: "DEBUG|process" "INFO|*" "ERROR|process/sync" and so on
```

Comment:
Don't we want to be able to set multiple packages loglevels? --log_level=”state:info,mempool:error,*:error” 

Found the idea here, don't know how actual this is: https://blog.cosmos.network/one-of-the-exciting-new-features-in-0-10-0-release-is-smart-log-level-flag-e2506b4ab756

Reply:
Right, good suggestion. Will modify

Reply:
Done, changed to the above mentioned pattern.